### PR TITLE
planner: Introduce historical stats tables

### LINKED ISSUE #18745: Support to export historical statistics
## Feature Request

**Is your feature request related to a problem? Please describe:**

There are some cases that when we(developer) get the wrong execution plan, statistics have been updated by `auto-analyze` or users without intention, which makes it harder to reproduce the plan and locate the reason.

**Describe the feature you'd like:**

Support to export historical statistics.
There is already an [API](https://docs.pingcap.com/tidb/stable/statistics#export-statistics) that can do this, but it has a strict limitation that the specified time cannot be earlier `tikv_gc_run_interval` minutes than the current time. (The default value of `tikv_gc_run_interval` is 10 minutes)
This limitation makes this API hard to be used in realistic cases.

It's better to design a new way to implement this feature.

**Teachability, Documentation, Adoption, Migration Strategy:**

[Export statistics in TiDB](https://docs.pingcap.com/tidb/stable/statistics#export-statistics)

### PR DESCRIPTION
**Pull request by An-DJ**
### What problem does this PR solve?

Issue Number: #18745 

Problem Summary:
Introduce 2 internal tables (mysql.stats_history and mysql.stats_meta_history) to store historical statistics

### What is changed and how it works?
Add 2 new tables to store historical statistics:
* `mysql.stats_history`
* `mysql.stats_meta_history`

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: session/bootstrap.go [hunk 1]
```diff
@@ -1714,6 +1732,10 @@ func doDDLWorks(s Session) {
 	mustExecute(s, CreateColumnStatsUsageTable)
 	// Create table_cache_meta table.
 	mustExecute(s, CreateTableCacheMetaTable)
+	// Create stats_history table.
+	mustExecute(s, CreateStatsHistory)
+	// Create stats_meta_history table.
+	mustExecute(s, CreateStatsMetaHistory)
```
Comment (qw4990) [inline session/bootstrap.go:1738]: We should create a new `ver` for them, here is an example https://github.com/pingcap/tidb/pull/26294/files, you can see changes of `bootstrap.go` in this PR.

File: session/bootstrap.go [hunk 2]
```diff
@@ -625,6 +645,7 @@ var (
 		upgradeToVer77,
 		upgradeToVer78,
 		upgradeToVer79,
+		upgradeToVer80,
```
Comment (qw4990) [inline session/bootstrap.go:648]: Please add some tests for it, you can refer to TestUpdateBindInfo

File: executor/infoschema_reader_test.go [hunk 3]
```diff
@@ -918,7 +918,7 @@ func (s *testInfoschemaClusterTableSuite) TestTableStorageStats(c *C) {
 	tk.MustQuery("select TABLE_SCHEMA, sum(TABLE_SIZE) from information_schema.TABLE_STORAGE_STATS where TABLE_SCHEMA = 'test' group by TABLE_SCHEMA;").Check(testkit.Rows(
 		"test 2",
 	))
-	c.Assert(len(tk.MustQuery("select TABLE_NAME from information_schema.TABLE_STORAGE_STATS where TABLE_SCHEMA = 'mysql';").Rows()), Equals, 27)
+	c.Assert(len(tk.MustQuery("select TABLE_NAME from information_schema.TABLE_STORAGE_STATS where TABLE_SCHEMA = 'mysql';").Rows()), Equals, 29)
```
Comment (qw4990) [inline executor/infoschema_reader_test.go:921]: Why did it change?

File: session/bootstrap_serial_test.go [hunk 4]
```diff
@@ -827,6 +827,45 @@ func TestUpgradeVersion75(t *testing.T) {
 	require.Equal(t, "char(255)", strings.ToLower(row.GetString(1)))
 }
 
+func TestUpgradeVersion80(t *testing.T) {
```
Comment (qw4990) [inline session/bootstrap_serial_test.go:830]: How about Implementing this test in this way: 1) upgrade to version 79; 2) check and ensure `stats_history` and `stats_meta_history` are not existed; 3) upgrade to the version 80; 4) check and ensure these two tables are existed;

File: executor/infoschema_reader_test.go (hunk 3 above)
Comment (qw4990) [reply executor/infoschema_reader_test.go:921]: It seems like it should not change.

File: session/bootstrap_serial_test.go (hunk 4 above)
Comment (An-DJ) [reply session/bootstrap_serial_test.go:830]: Okk. I considered that the `stats_history` and `stats_meta_history` could be created manually at version 79. That could affect the upgrade progress so I dropped it if existed. 

It seems that checking it before upgrade is more reasonable. I will take it.

File: executor/infoschema_reader_test.go (hunk 3 above)
Comment (An-DJ) [reply executor/infoschema_reader_test.go:921]: I guess it checks the tables info in `mysql` database. More details can be found in #30066.

File: session/bootstrap.go [hunk 5]
```diff
@@ -365,6 +365,24 @@ const (
 		oldReadLease bigint(20) NOT NULL DEFAULT 0,
 		PRIMARY KEY (tid)
 	);`
+	// CreateStatsHistory stores the historical stats.
+	CreateStatsHistory = `CREATE TABLE IF NOT EXISTS mysql.stats_history (
+		table_id bigint(64) NOT NULL,
+		stats_data longblob NOT NULL,
+		seq_no bigint(64) NOT NULL,
+		version bigint(64) NOT NULL,
+		create_time datetime NOT NULL,
```
Comment (qw4990) [inline session/bootstrap.go:374]: Please use datetime with fraction `create_time datetime(6)`.

File: session/bootstrap.go [hunk 6]
```diff
@@ -365,6 +365,24 @@ const (
...
+		PRIMARY KEY(table_id)
+	);`
+	// CreateStatsMetaHistory stores the meta historical stats.
+	CreateStatsMetaHistory = `CREATE TABLE IF NOT EXISTS mysql.stats_meta_history (
+		table_id bigint(64) NOT NULL,
+		modify_count bigint(64) NOT NULL,
+		count bigint(64) NOT NULL,
+		version bigint(64) NOT NULL,
+		create_time datetime NOT NULL,
```
Comment (qw4990) [inline session/bootstrap.go:383]: Please use datetime with fraction create_time datetime(6).

Comment (An-DJ) [reply session/bootstrap.go]: Done

**Review by sekirooooo — APPROVED** [review]

**Review by rebelice — APPROVED** [review]

**Review by Reminiscent — APPROVED** [review]

**Comment by Reminiscent** [comment]
@An-DJ Please solve conflicts

File: executor/infoschema_reader_test.go [hunk 7]
```diff
@@ -17,6 +17,7 @@ package executor_test
 import (
 	"crypto/tls"
 	"fmt"
+	"github.com/pingcap/tidb/path/pkg/mod/github.com/go-kit/kit@v0.9.0/transport/grpc"
```
Comment (Reminiscent) [inline executor/infoschema_reader_test.go:20]: Please format this.

**Comment by Mini256** [comment]
@qw4990，PTAL again~ please explicitly approve to indicate that the PR allows merging.

**Review by wjhuang2016 — COMMENTED** [review]
Since it modify the meta, it would be strange for users to upgrade TiDB version and found creating two unused tables.
So I suggest you to develop this feature in a seperated branch. After you finish it, and we can merge it together.

**Comment by An-DJ** [comment]
> Since it modify the meta, it would be strange for users to upgrade TiDB version and found creating two unused tables. So I suggest you to develop this feature in a seperated branch. After you finish it, and we can merge it together.

@wjhuang2016 This 2 tables creation is part of #18745. We want to use them to store hsitorical statistics info of TiDB. The following works are based on this PR. PTAL~

**Comment by rebelice** [comment]
We created a new [pull/31032](https://github.com/pingcap/tidb/pull/31032) to replace this one.

**Pull request closed without merging**